### PR TITLE
Changes in the muon block

### PR DESCRIPTION
**Pull request by MRD2F**

### CODE REVIEW

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 1]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+
+struct MuonHitMatchV2 {
+
+    static constexpr size_t n_muon_stations = 4;
+    static constexpr int first_station_id = 1;
+    static constexpr int last_station_id = first_station_id + n_muon_stations - 1;
+    using CountArray = std::array<unsigned, n_muon_stations>;
+    using CountMap = std::map<int, CountArray>;
+
+    const std::vector<int>& ConsideredSubdets()
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 2]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+    {
+        static const std::vector<int> subdets = { MuonSubdetId::DT, MuonSubdetId::CSC, MuonSubdetId::RPC };
+        return subdets;
+    }
+
+    const std::string& SubdetName(int subdet)
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 3]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+    {
+        static const std::map<int, std::string> subdet_names = {
+            { MuonSubdetId::DT, "DT" }, { MuonSubdetId::CSC, "CSC" }, { MuonSubdetId::RPC, "RPC" }
+        };
+        if(!subdet_names.count(subdet))
+            throw cms::Exception("MuonHitMatch") << "Subdet name for subdet id " << subdet << " not found.";
+        return subdet_names.at(subdet);
+    }
+
+    size_t GetStationIndex(int station, bool throw_exception) const
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 4]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+        for(int subdet : ConsideredSubdets()) {
+            n_matches[subdet].fill(0);
+            n_hits[subdet].fill(0);
+        }
+
+        CountMatches(muon, n_matches);
+        CountHits(muon, n_hits);
+    }
+
+    void CountMatches(const pat::Muon& muon, CountMap& n_matches)
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 5]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+        for(const auto& segment : muon.matches()) {
+            if(segment.segmentMatches.empty() && segment.rpcMatches.empty()) continue;
+            if(n_matches.count(segment.detector())) {
+                const size_t station_index = GetStationIndex(segment.station(), true);
+                ++n_matches.at(segment.detector()).at(station_index);
+            }
+        }
+    }
+
+    void CountHits(const pat::Muon& muon, CountMap& n_hits)
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 6]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+
+                        if(muon_n_hits)
+                            ++muon_n_hits->at(station_index);
+                    }
+                }
+            }
+        }
+    }
+
+    unsigned NMatches(int subdet, int station) const
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 7]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+    {
+        if(!n_matches.count(subdet))
+            throw cms::Exception("MuonHitMatch") << "Subdet " << subdet << " not found.";
+        const size_t station_index = GetStationIndex(station, true);
+        return n_matches.at(subdet).at(station_index);
+    }
+
+    unsigned NHits(int subdet, int station) const
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 8]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+    {
+        if(!n_hits.count(subdet))
+            throw cms::Exception("MuonHitMatch") << "Subdet " << subdet << " not found.";
+        const size_t station_index = GetStationIndex(station, true);
+        return n_hits.at(subdet).at(station_index);
+    }
+
+    unsigned CountMuonStationsWithMatches(int first_station, int last_station) const
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 9]
```diff
@@ -294,6 +307,147 @@ struct MuonHitMatchV1 {
...
+        unsigned cnt = 0;
+        for(size_t n = first_station_index; n <= last_station_index; ++n) {
+            for(const auto& match : n_matches) {
+                if(!masks.at(match.first).at(n) && match.second.at(n) > 0) ++cnt;
+            }
+        }
+        return cnt;
+    }
+
+    unsigned CountMuonStationsWithHits(int first_station, int last_station) const
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: method name should start with a lowercase letter

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 10]
```diff
@@ -881,11 +1035,141 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+                get(pfCand_muon_vertex_dz_tauFL) = getValueNorm(pfCands.at(index_pf_muon).vertex().z() -
+                    pv.position().z() - tau.flightLength().z(), -0.0118f, 4.405f);
+
+                const bool hasTrackDetails = pfCands.at(index_pf_muon).hasTrackDetails();
+                if(hasTrackDetails){
+                    get(pfCand_muon_hasTrackDetails) = hasTrackDetails;
+                    get(pfCand_muon_dxy) = getValueNorm(std::abs(pfCands.at(index_pf_muon).dxy()), -0.0045f, 0.9655f);
+                    get(pfCand_muon_dxy_sig) = getValueNorm(std::abs(pfCands.at(index_pf_muon).dxy()) /
+                        pfCands.at(index_pf_muon).dxyError(), 4.575f, 42.36f);
+                    get(pfCand_muon_dz) =  getValueNorm(std::abs(pfCands.at(index_pf_muon).dz()), -0.0117f, 4.097f);
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: dz shouldn't be abs

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 11]
```diff
@@ -881,11 +1035,141 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+
+                get(muon_valid) = valid_index_muon;
+                get(muon_rel_pt) = getValueNorm(muons.at(index_muon).polarP4().pt() / tau.polarP4().pt(),
+                    is_inner ?  0.7966f : 0.2678f, is_inner ? 3.402f : 3.592f);
+                get(muon_deta) = getValueLinear(muons.at(index_muon).polarP4().eta() - tau.polarP4().eta(),
+                    is_inner ? -0.1f : -0.5f, is_inner ? 0.1f : 0.5f, false);
+                get(muon_dphi) = getValueLinear(dPhi(tau.polarP4(), muons.at(index_muon).polarP4()),
+                    is_inner ? -0.1f : -0.5f, is_inner ? 0.1f : 0.5f, false);
+                get(muon_dxy) = getValueNorm(muons.at(index_muon).dB(pat::Muon::PV2D), 0.0019f, 1.039f);
+                get(muon_dxy_sig) = getValueNorm(muons.at(index_muon).dB(pat::Muon::PV2D) /
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: dB should be abs in sig estimation.

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 12]
```diff
@@ -881,11 +1035,141 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+                    muons.at(index_muon).edB(pat::Muon::PV2D), 8.98f, 71.17f);
+
+                const bool normalizedChi2_valid = muons.at(index_muon).globalTrack().isNonnull() && muons.at(index_muon).normChi2();
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: ... normChi2() >= 0

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 13]
```diff
@@ -881,11 +1035,141 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+                if(normalizedChi2_valid){
+                    get(muon_normalizedChi2_valid) = normalizedChi2_valid;
+                    get(muon_normalizedChi2) = getValueNorm(muons.at(index_muon).normChi2(), 21.52f, 265.8f);
+                    if(muons.at(index_muon).innerTrack().isNonnull())
+                        get(muon_numberOfValidHits) = getValueNorm(muons.at(index_muon).numberOfValidHits(), 21.84f, 10.59f);
+                }
+                get(muon_segmentCompatibility) = getValue(muons.at(index_muon).segmentCompatibility());
+                get(muon_caloCompatibility) = getValue(muons.at(index_muon).caloCompatibility());
+
+                const bool pfEcalEnergy_valid = muons.at(index_muon).pfEcalEnergy();
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: ... pfEcalEnergy() >= 0

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 11 above)
Comment (MRD2F) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: like this: 
```
getValueNorm(std::abs(muons.at(index_muon).dB(pat::Muon::PV2D)) / 
           muons.at(index_muon).edB(pat::Muon::PV2D), 8.98f, 71.17f);
```
?

Comment (kandrosov) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: yes

**Review by kandrosov — APPROVED** [review]

**Pull request merged**